### PR TITLE
Tag axes of discretization-related arrays

### PR DESCRIPTION
**Pull request by inducer**
This takes #188 (in the form of https://github.com/kaushikcfd/grudge/commit/3d184d5e6886574cce2fb596e84091272d92a60e) and modernizes it for the merged form of https://github.com/inducer/meshmode/pull/284, and adds two commits to help avoid things that are now deprecated. This also includes https://github.com/kaushikcfd/grudge/commit/1d89eb28637ee89b6d3722d43639e029fa546d1e, because why not.

Empirically, with [this hacked version](https://github.com/inducer/meshmode/compare/kaushik-hacked) of https://github.com/kaushikcfd/meshmode/commits/main, it is still capable of transforming mirgecom code.

Best read commit-by-commit.

### CODE REVIEW

File: grudge/array_context.py [hunk 1]
```diff
@@ -250,6 +284,11 @@ def __call__(self, arg_id_to_arg) -> ArrayContainer:
         from arraycontext.impl.pytato.utils import get_cl_axes_from_pt_axes
         input_args_for_prg = _args_to_device_buffers(
                 self.actx, self.input_id_to_name_in_program, arg_id_to_arg)
+        from time import time
+
+        self.actx.queue.finish()
+
+        t_start = time()
```
Comment (kaushikcfd) [inline grudge/array_context.py:291]:
```suggestion
````

File: grudge/array_context.py [hunk 2]
```diff
@@ -258,6 +297,10 @@ def __call__(self, arg_id_to_arg) -> ArrayContainer:
                 allocator=self.actx.allocator,
                 input_args=input_args_for_prg)
 
+        t_end = time()
+
+        print(f"Fusion actx took: {t_end - t_start} secs")
+
```
Comment (kaushikcfd) [inline grudge/array_context.py:303]:
```suggestion
```
Sorry, this was a spurious diff that had sprung up into my tree at some point.

File: grudge/trace_pair.py [hunk 3]
```diff
@@ -419,7 +419,8 @@ def communicate_single_array(key, local_bdry_ary):
                     local_bdry_ary, dest_rank=remote_rank, comm_tag=ary_tag,
                     stapled_to=make_distributed_recv(
                         src_rank=remote_rank, comm_tag=ary_tag,
-                        shape=local_bdry_ary.shape, dtype=local_bdry_ary.dtype))
+                        shape=local_bdry_ary.shape, dtype=local_bdry_ary.dtype,
+                        axes=local_bdry_ary.axes))
```
Comment (kaushikcfd) [inline grudge/trace_pair.py:423]: Needs https://github.com/inducer/pytato/pull/339

**Review by kaushikcfd — APPROVED** [review]
Thanks!

File: grudge/trace_pair.py (hunk 3 above)
Comment (inducer) [reply grudge/trace_pair.py:423]: Thanks! That's in now, so let's see how we look.

File: grudge/array_context.py (hunk 2 above)
Comment (inducer) [reply grudge/array_context.py]: Thanks, removed.

File: grudge/array_context.py (hunk 1 above)
Comment (inducer) [reply grudge/array_context.py]: Thanks, removed!

**Comment by inducer** [comment]
Thanks for taking a look, and for working on this in the first place! Took a final look, LGTM. Setting to merge, assuming it passes.

**Pull request merged**